Implement user authorization

ashlynbaum opened the pull request:
# Implementation progress
- [x] Registration - plain text passwords
- [x] Login - `(email, password) => id`
- [x] Encrypt password 
- [x] Authorize users based on `id` - secure resources
- [x] Check email validity
- [x] Authentication based on generated secure tokens
- [x] Basic Access Authentication & `Authorization` header
- [x] HTTPS!
- [x] Race condition on token generation

![XKCD Security](http://imgs.xkcd.com/comics/security.png)
https://xkcd.com/538/
# Definitions

Three importants definitons taken from https://stackoverflow.com/questions/6556522/authentication-versus-authorization
- **Identification** is claiming you are somebody
- **Authentication** is the process of ascertaining that somebody really is who he claims to be.
- **Authorization** refers to rules that determine who is allowed to do what. E.g. Adam may be authorized to create and delete databases, while Usama is only authorised to read.
# Implementation steps
## Registration - plain text passwords

Endpoint that receive `email` and `password` and store them in the database.
## Login - returns id as authentication token

Just a simple endpoint that that returns the `id` of a user **identified** by his `email` and then **authenticated** by his password.

The reason why we return the `id` here is because we are going to assume that the `id` is secret enought that you can authorize someone to create an event or modify an event by just providing his user `id` and the `creatorId` of an event.
## Authorize users based on `id` - secure resources

We are now going to secure our events CRUD. The reading actions ( list events and show an event ) are public but the other are not.
At the moment we using the `id` of the user as his authentication token and pass the token as a `X-Auth-Token` header for authorized requests.

These simple authorizations define how we protect our event resource :
- Any unauthenticated user can read events
- Any authenticated user can create an event
- Update an event requires to be the `creator` of the event
- Delete an event requires to be the `creator` of the event

This implementation can be splited in a few step itself:
- [x] Attach events to users when events are created
- [x] Authenticate users based on his authentication token for each secure endpoint
- [x] Implement authorizations

**Don't forget to hide the `creatorId` of events since it's the secret use to identify and authenticate them !**
## Check email validity

At the moment it's possible with "@WoOoT@" as a valid email address and that's an issue, especialy if we want to confirm email adresses later. The [SMTP server](https://en.wikipedia.org/wiki/Simple_Mail_Transfer_Protocol) is going to be very angry against us.

I think you already found a regex for this, so just use it and we're going to analyze it together later.
## Authentication based on generated secure tokens

User `id` is not the best way to authenticate a user for a frew reasons. We are going to make userId public, to be able to request things such as `GET /users/{id}` to see a profile and also because it's easy to predict [MongoDB ObjectID](http://docs.mongodb.org/manual/reference/object-id/).

We are going to generate random strings such as `7704853336302830aaa609584bb1fa426a00c6331a47e13851b710b4e18a42f4` to authenticate our users.
Because a computer is **deterministic**, it's impossible to generate randomness with a programm. To generate "random numbers", entropy comes from, temperature captors, electronic signal noise, mouse mouvements, everything that's difficult to predict is part of the equasion...

node.js strandard library comes with a [`crypto`](https://nodejs.org/api/crypto.html) package that have a [`crypto.randomBytes` function](https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback).
There is a comment about the fact that the function call might fail. Our strategy is to wait for using [`setTimeout`](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setTimeout) and retry generating the random number.
## Basic Access Authentication & `Authorization` header

Using a header named `X-Auth-Token` to sent the token and passing user credentials as a json playload in the `POST /login` was very simple way to transfert those data. Unfortunatly none of those are very standart even if there are 

Like any other header, the `Authorization` one is just a string value sent between the server and the client but there are a few conventions around this header.
The `Authorization` header is part of the [RFC1945](https://tools.ietf.org/html/rfc1945) that define the HTTP/1.0 protocol... [Section 10.2](https://tools.ietf.org/html/rfc1945#section-10.2) and [Section 11(https://tools.ietf.org/html/rfc1945#section-11). You'll find in these two sections everything you need to change the `POST /login` endpoint.
### Basic Scheme

So, instead of sending `email` and `password` through a json playload, we are going to use the [Authorization Basic Scheme](https://tools.ietf.org/html/rfc1945#section-11.1).
The Basic Sheme requires to encode and decode using the [standard base64 encoding](https://en.wikipedia.org/wiki/Base64) you can play with base64 encoding using `openssl` as in the following examples, for node.js you can either use Javascript `btoa` and `atob` functions or use node's [Buffer#toString](https://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end) method.

``` bash
# To encode
$ echo "encode this" | openssl enc -base64
ZW5jb2RlIHRoaXMK

# To decode
$ echo "ZW5jb2RlIHRoaXMK" | openssl enc -d -base64
encode this
```

So at the end, a request like this should return a user authentication token:

```
GET /
Authorization: Basic PDMgQXNobHluCg==
```

_Notice that I expect to `GET /` and not `POST /login`._
### Token Scheme

We are now following a nice convention for our Authentication process but our Authorization token is still sent in the `X-Auth-Token` header of authenticated requests. We can define our own scheme that's not the Basic one for the purpose of sending the authentication token.

Here is the scheme I suggest to use:

```
basic-credentials = "Token" SP token
token             = <user token string>
```
## HTTPS!

Last but not least, I want the server to respond `426 Upgrade Required` for any non HTTPS request. No content, don't do anything, just catch all non HTTPS requests and send a `426` status.

Unfortunatly there is no way to disable the `80` port on heroku...
## Race condition on token generation

When you send multiple http requests to a node.js server, those requests are processes concurently, which mean that requests are processes about at the same time.

``` javascript
var genUniqueToken = function(cb) {
  var token = genToken();
  isTokenInDatabase(function(isPresent) {
    if(isPresent) {
      genUniqueToken(cb);
    } else {
      cb(null, token);
    }
  })
};

genUniqueToken(function(err, token) {
  // t1
  insertInDatabase({token: token}, function(err, result) {
    // t2
  });
});
```

If you have two request executing concurently the previous code with `t1` < `t1'` < `t2` < `t2'` ( `t1` is the time when the code following the comment `t1` is exectuted ) and by an impossible hazard the two request have generated the same token, then you'll have two documents with the same token in your database... because both requests checked the existance of the token before the other inserted the new document.

To ensure that it's not possible to have two document with the same token, you'll have to use a [mongodb index](http://docs.mongodb.org/manual/core/indexes-introduction/).

[`Collection#ensureIndex`](https://mongodb.github.io/node-mongodb-native/markdown-docs/indexes.html#ensure-indexes-with-ensureindex) is what you need to create an index.

In case of index violation, a call to collection.insert will give you back an error...

Commit:
Implement user authorization prototype

Update API to account for user authorizion. Add user sign in
and user login functionality. Store passwords as hash in database.

barodeur commented:
I think you forgot to add `bcrypt` as a dependency

barodeur commented:
You need to check that the email is a correct email

Commit:
Add bcrypt as a dependancy

barodeur commented:
Then to authenticate the others request ( such as `POST /events` ) we are going to send credentials.
But we are not gonna send `(email,password)` but an `authenticationToken`, which is a **unique** string generated for each user.
Can you modify the login so it return this `authenticationToken`. A good time to generate this token is at signup.

Commit:
Verify email address for user sign up

barodeur commented:
1. When posing a new event, find how to read the `X-Auth-Token` http header of the `POST /events` and store it as the attribute `userAuthToken` of the new event.
2. generate a unique token ( two users cann't have the same `authToken`) for each new user signing up. Let' s say the token generated is a 32 char long hexadecimal string.
3. Change you signup so it returns the `authToken`
4. Authenticate the user based on the `X-Auth-Token` header when posting a new event `POST /events` and instead of adding the `userAuthToken`, add the `userId` to the event.
5. Implement an authorization check so only the creator of an event can modify or delete his event.
6. Use Basic Auth header istead of `X-Auth-Token`

barodeur commented:
@ashlynbaum here is a gist that might help you with the token generation. You just have to replace the mocked database call https://gist.github.com/barodeur/475fc14953acd0a398ff

Commit:
Refactor email validation

barodeur commented:
@ashlynbaum Just added a new oppotrunity for you to play with asynchronous rescusive functions :wink: I named it "Race condition on token generation" :sparkles:

Commit:
Authenticate users for event creation

Attatch user ID as creatorID to events when they are created.
The User Id is temporary, and will be replaced with authentication token.
The creatorID is authenticated by checking if the ID matches any user.

Commit:
Implement user authorization for events

Authentication: users must be creators of event to patch or delete event.
Middleware is used for identification and authentication.

Commit:
Generate unique token and assign to user

Commit:
fix syntax

Commit:
fix style and syntax

Commit:
Genrate random authentication token

Commit:
Update authentication middleware

update authentication middleware to use randomly generated
authentication token instead of user ID

Commit:
fixed styling for uniformity

Commit:
Implement basic authorization scheme for login

Commit:
Fix syntax for dependancies

Commit:
Continue Auth

set up authorization middleware to recieve authorization
tokens from the authorization header.

Commit:
Block non HTTPS requests in production

Commit:
Remove unnecessary log

Commit:
Fix bud: Server always respond upgrade to HTTPS

Read X-Forward-proto header to check for HTTPS connection.

Commit:
Fix bug: crash when authorization header is not sent

Commit:
Display port server is listening to

Commit:
Use Index to ensure user token is unique

Refactor the generation of unique token to automatically insert user into
database to fix race condition bug.

Commit:
Refactor the generation of authentication token

Commit:
merge master into branch, fix conflict
